File: src/js/lib/canonicaliser.js
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

import {
	AstAlternation,
	AstChars,
	AstDigitsRange,
	AstEmpty,
	AstInterval,
	AstOptional,
	AstReplaceable,
	AstSequence,
} from './ast.js';
import { SingleStringOutcome, advance, tryGetSingleString } from './matcher.js';
import { NaxpLimits } from './naxp-limits.js';

/**
 * Computes ρ, the map from an accepted string to its canonical form.
 *
 * ρ(*w*) is *w* with the match of each replaceable element replaced by that element's rendering.
 * The tree is where that is visible, since the machines have already resolved it one way or the
 * other, so this works over the tree.
 *
 * It is the matcher's set of positions with one output carried alongside each position. A
 * replaceable element contributes its rendering whatever it matched, which is the whole of what
 * makes the canonical form differ from the input.
 *
 * One output per position is enough, and that rests on W3 being decided when the naxp was
 * compiled. Everything reached at a given point in the walk has the same future, since what
 * follows depends on the position alone; so two partial parses that meet at one position either
 * both reach the end or neither does. If both reach it they append the same remainder, and W3 says
 * the two totals agree, which forces the two outputs to have agreed already. Carrying the whole set
 * would therefore only ever record the same string twice — and it was what made this exponential,
 * since `([ab]|[ab]!a){17}` reaches 2^17 outputs on an all-`b` input.
 */
class Canonicaliser {
	/**
	 * @param {string} text The string being canonicalised.
	 */
	constructor(text) {
		this.text = text;
		/** @type {Map<AstReplaceable, string>} */
		this.renderings = new Map();
	}

	/**
	 * @param {import('./ast.js').Ast} node The node.
	 * @param {Map<number, string>} starts The positions reached so far, with their outputs.
	 * @returns {Map<number, string>} The positions reached.
	 */
	advance(node, starts) {
		if (starts.size === 0) { return starts; }

		if (node instanceof AstEmpty) { return starts; }

		if (node instanceof AstChars) {
			const result = new Map();

			for (const [at, output] of starts) {
				if (at >= this.text.length) { continue; }

				if (node.charSet.contains(this.text.charCodeAt(at))) {
					put(result, at + 1, output + this.text[at]);
				}
			}

			return result;
		}

		// A digits range emits what it consumed.
		if (node instanceof AstDigitsRange) { return this.consume(node, starts); }

		if (node instanceof AstSequence) {
			let current = starts;

			for (const child of node.children) {
				current = this.advance(child, current);

				if (current.size === 0) { break; }
			}

			return current;
		}

		if (node instanceof AstAlternation) {
			const result = new Map();

			for (const child of node.children) { putAll(result, this.advance(child, starts)); }

			return result;
		}

		if (node instanceof AstOptional) {
			const result = new Map(starts);

			putAll(result, this.advance(node.child, starts));

			return result;
		}

		if (node instanceof AstInterval) {
			const result = node.minCount === 0 ? new Map(starts) : new Map();
			let current = starts;

			for (let i = 1; i <= node.maxCount; ++i) {
				const next = this.advance(node.child, current);

				if (next.size === 0) { break; }

				if (i >= node.minCount) { putAll(result, next); }

				// A child that matches nothing reaches a fixed point at once.
				if (i >= node.minCount && sameAs(next, current)) { break; }

				current = next;
			}

			return result;
		}

		if (node instanceof AstReplaceable) {
			// This is the whole of the difference between a string and its canonical form:
			// whatever the subject matched, the rendering is what comes out.
			const rendering = this.renderingOf(node);
			const result = new Map();

			for (const [at, output] of starts) {
				for (const end of advance(node.subject, this.text, new Set([at]))) {
					put(result, end, output + rendering);
				}
			}

			return result;
		}

		throw new Error(`Unhandled node type ${node.constructor.name}.`);
	}

	/**
	 * Advances by a node that emits exactly the characters it consumed.
	 *
	 * @param {import('./ast.js').Ast} node The node.
	 * @param {Map<number, string>} starts The positions.
	 * @returns {Map<number, string>} The positions reached.
	 */
	consume(node, starts) {
		const result = new Map();

		for (const [at, output] of starts) {
			for (const end of advance(node, this.text, new Set([at]))) {
				put(result, end, output + this.text.slice(at, end));
			}
		}

		return result;
	}

	/**
	 * @param {AstReplaceable} replaceable The element.
	 * @returns {string} Its rendering.
	 */
	renderingOf(replaceable) {
		const cached = this.renderings.get(replaceable);

		if (cached !== undefined) { return cached; }

		// W1 has already established that the rendering generates exactly one string.
		const { outcome, result } = tryGetSingleString(replaceable.rendering);

		if (outcome !== SingleStringOutcome.Single) {
			throw new Error('A replaceable element passed W1 but has no single rendering.');
		}

		this.renderings.set(replaceable, result);

		return result;
	}
}

/**
 * Records an output for a position, keeping whichever arrived first.
 *
 * Under W3 a second one that matters cannot differ from the first; see the note on the class.
 *
 * @param {Map<number, string>} target Where to record it.
 * @param {number} end The position.
 * @param {string} output What was emitted to reach it.
 */
function put(target, end, output) {
	if (!target.has(end)) { target.set(end, output); }
}

/**
 * @param {Map<number, string>} target Where to record them.
 * @param {Map<number, string>} source What to record.
 */
function putAll(target, source) {
	for (const [end, output] of source) { put(target, end, output); }
}

/**
 * @param {Map<number, string>} left The first.
 * @param {Map<number, string>} right The second.
 * @returns {boolean} Whether they agree on every position and output.
 */
function sameAs(left, right) {
	if (left.size !== right.size) { return false; }

	for (const [end, output] of left) {
		if (right.get(end) !== output) { return false; }
	}

	return true;
}

/**
 * The canonical form of a string.
 *
 * @param {import('./ast.js').Ast} ast The parsed naxp, which must have been through the W3
 * checker.
 * @param {string} text The string.
 * @returns {string | null} The canonical form, or null where the naxp does not accept the string.
 */
export function tryCanonicalise(ast, text) {
	if (ast === null || ast === undefined) { throw new TypeError('ast is required.'); }

	// A naxp within the state budget has a longest string shorter than this, so anything longer
	// is not accepted rather than too costly to decide.
	if (text.length > NaxpLimits.maxStringLength) { return null; }

	const reached = new Canonicaliser(text).advance(ast, new Map([[0, '']]));
	const output = reached.get(text.length);

	return output === undefined ? null : output;
}

File: src/cs/Naxp/Canonicaliser.cs
// Copyright (c) Tim Gordon.
// This file is licensed to you under the Apache Licence, Version 2.0. See the LICENSE file.

using System;
using System.Collections.Generic;

namespace LogMu;

/// <summary>
/// Computes &#961;, the map from an accepted string to its canonical form.
/// </summary>
/// <remarks>
/// <para>
/// &#961;(<i>w</i>) is <i>w</i> with the match of each replaceable element replaced by that
/// element's rendering. The tree is where that is visible, since the machines have already
/// resolved it one way or the other, so this works over the tree.
/// </para>
/// <para>
/// It is <see cref="Matcher"/>'s set of positions with one output carried alongside each
/// position. A replaceable element contributes its rendering whatever it matched, which is the
/// whole of what makes the canonical form differ from the input.
/// </para>
/// <para>
/// One output per position is enough, and that rests on W3 being decided when the naxp was
/// compiled. Everything reached at a given point in the walk has the same future, since what
/// follows depends on the position alone; so two partial parses that meet at one position either
/// both reach the end or neither does. If both reach it they append the same remainder, and W3
/// says the two totals agree, which forces the two outputs to have agreed already. Carrying the
/// whole set would therefore only ever record the same string twice — and it was what made this
/// exponential, since <c>([ab]|[ab]!a){17}</c> reaches 2^17 outputs on an all-<c>b</c> input.
/// </para>
/// </remarks>
readonly ref struct Canonicaliser
{
	readonly ReadOnlySpan<char> text;
	readonly Dictionary<Ast, string> renderings;

	Canonicaliser(ReadOnlySpan<char> text)
	{
		this.text = text;
		this.renderings = [];
	}

	/// <summary>
	/// The canonical form of a string.
	/// </summary>
	/// <param name="ast">The parsed naxp, which must have been through <see cref="W3Checker"/>.</param>
	/// <param name="text">The string.</param>
	/// <param name="canonical">The canonical form, or <see langword="null"/> where there is none.</param>
	/// <returns>Whether the naxp accepts the string.</returns>
	public static bool TryCanonicalise(Ast ast, ReadOnlySpan<char> text, out string? canonical)
	{
		if (ast is null) { throw new ArgumentNullException(nameof(ast)); }

		canonical = null;

		// A naxp within the state budget has a longest string shorter than this, so anything
		// longer is not accepted rather than too costly to decide.
		if (text.Length > NaxpLimits.MaxStringLength) { return false; }

		var canonicaliser = new Canonicaliser(text);
		Dictionary<int, string> reached = canonicaliser.Advance(
			ast,
			new Dictionary<int, string> { [0] = string.Empty });

		if (!reached.TryGetValue(text.Length, out string? output)) { return false; }

		canonical = output;
		return true;
	}

	Dictionary<int, string> Advance(Ast node, Dictionary<int, string> starts)
	{
		if (starts.Count == 0) { return starts; }

		switch (node)
		{
			case AstEmpty:
				return starts;

			case AstChars chars:
			{
				var result = new Dictionary<int, string>();

				foreach (KeyValuePair<int, string> start in starts)
				{
					if (start.Key < this.text.Length && chars.CharSet.Contains(this.text[start.Key]))
					{
						Put(result, start.Key + 1, start.Value + this.text[start.Key]);
					}
				}

				return result;
			}

			case AstDigitsRange:
				// A digits range emits what it consumed.
				return this.Consume(node, starts);

			case AstSequence sequence:
			{
				var current = starts;

				foreach (Ast child in sequence.Children)
				{
					current = this.Advance(child, current);
					if (current.Count == 0) { break; }
				}

				return current;
			}

			case AstAlternation alternation:
			{
				var result = new Dictionary<int, string>();
				foreach (Ast child in alternation.Children) { PutAll(result, this.Advance(child, starts)); }

				return result;
			}

			case AstOptional optional:
			{
				var result = new Dictionary<int, string>(starts);
				PutAll(result, this.Advance(optional.Child, starts));

				return result;
			}

			case AstInterval interval:
			{
				Dictionary<int, string> result = interval.MinCount == 0
					? new Dictionary<int, string>(starts)
					: new Dictionary<int, string>()
					;
				Dictionary<int, string> current = starts;

				for (int i = 1; i <= interval.MaxCount; ++i)
				{
					Dictionary<int, string> next = this.Advance(interval.Child, current);
					if (next.Count == 0) { break; }

					if (i >= interval.MinCount) { PutAll(result, next); }

					// A child that matches nothing reaches a fixed point at once.
					if (SameAs(next, current) && i >= interval.MinCount) { break; }

					current = next;
				}

				return result;
			}

			case AstReplaceable replaceable:
			{
				// This is the whole of the difference between a string and its canonical form:
				// whatever the subject matched, the rendering is what comes out.
				string rendering = this.RenderingOf(replaceable);
				var result = new Dictionary<int, string>();

				foreach (KeyValuePair<int, string> start in starts)
				{
					foreach (int end in Matcher.Advance(replaceable.Subject, this.text, new HashSet<int> { start.Key }))
					{
						Put(result, end, start.Value + rendering);
					}
				}

				return result;
			}

			default:
				throw new InvalidOperationException($"Unhandled node type {node.GetType().Name}.");
		}
	}

	/// <summary>
	/// Advances by a node that emits exactly the characters it consumed.
	/// </summary>
	Dictionary<int, string> Consume(Ast node, Dictionary<int, string> starts)
	{
		var result = new Dictionary<int, string>();

		foreach (KeyValuePair<int, string> start in starts)
		{
			foreach (int end in Matcher.Advance(node, this.text, new HashSet<int> { start.Key }))
			{
				Put(result, end, start.Value + this.text.Slice(start.Key, end - start.Key).ToString());
			}
		}

		return result;
	}

	string RenderingOf(AstReplaceable replaceable)
	{
		if (this.renderings.TryGetValue(replaceable, out string? cached)) { return cached; }

		// W1 has already established that the rendering generates exactly one string.
		if (Matcher.TryGetSingleString(replaceable.Rendering, out string? rendering) != SingleStringOutcome.Single)
		{
			throw new InvalidOperationException("A replaceable element passed W1 but has no single rendering.");
		}

		this.renderings.Add(replaceable, rendering!);

		return rendering!;
	}

	/// <summary>
	/// Records an output for a position, keeping whichever arrived first.
	/// </summary>
	/// <remarks>
	/// Under W3 a second one that matters cannot differ from the first; see the note on the class.
	/// </remarks>
	static void Put(Dictionary<int, string> target, int end, string output)
	{
		if (!target.ContainsKey(end)) { target.Add(end, output); }
	}

	static void PutAll(Dictionary<int, string> target, Dictionary<int, string> source)
	{
		foreach (KeyValuePair<int, string> item in source) { Put(target, item.Key, item.Value); }
	}

	static bool SameAs(Dictionary<int, string> left, Dictionary<int, string> right)
	{
		if (left.Count != right.Count) { return false; }

		foreach (KeyValuePair<int, string> item in left)
		{
			if (!right.TryGetValue(item.Key, out string? other)) { return false; }
			if (!string.Equals(item.Value, other, StringComparison.Ordinal)) { return false; }
		}

		return true;
	}
}
